### PR TITLE
Add mixed-effect HMM example on harbour seal data

### PR DESCRIPTION
**Pull request by eb8680**
This PR ports the seal experiment from the TVE paper to Funsor, including a version that uses `sequential_sum_product`.  I believe this is working correctly and I'm getting log-likelihood numbers and parameter values that make sense.  

Remaining tasks:
- [x] remove sequential model in favor of naive sequential sum product
- [x] tweak zero inflation logic to enumerate

### CODE REVIEW

File: examples/mixed_hmm/seal_data.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+import os
```
Comment (eb8680) [inline examples/mixed_hmm/seal_data.py]: This data-handling file is copied verbatim from the Pyro example.

**Comment by martinjankowiak** [comment]
took a quick look but i'm afraid this is too much code for me to be able to give a thorough CR right now.

your hypothesis about improved numerical stability certainly seems plausible to me. one thing you might do to convince yourself that that is indeed what is going on is clip all the observed sequences to some much shorter length (e.g. T=50) and rerun both versions. presumably there shouldn't be much room for deltas driven by numerical (in)stability in that regime.

File: examples/mixed_hmm/experiment.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        s = 1
+        for d in tensor.shape:
+            s = s * d
+        return s
+
+    with poutine.block(), poutine.trace(param_only=True) as param_capture:
+        model()
+        guide()
+
+    return sum(_size(node["value"]) for node in param_capture.trace.nodes.values())
```
Comment (fritzo) [inline examples/mixed_hmm/experiment.py]: nit: use [numel()](https://pytorch.org/docs/stable/tensors.html#torch.Tensor.numel), as in `node["value"].numel()`

File: examples/mixed_hmm/experiment.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+
+def sequential_loss_fn(model, guide):
+    # XXX ignore guide for now
```
Comment (fritzo) [inline examples/mixed_hmm/experiment.py]: Can we replace this comment with something more final?
```
# we perform exact inference, so the guide is trivial
```

File: examples/mixed_hmm/experiment.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    loss = apply_optimizer(loss)
+    assert not loss.inputs
+    return -loss.data
+
+
+def parallel_loss_fn(model, guide):
+    # XXX ignore guide for now
+    factors = model()
+    t_term, new_factors = factors[0], factors[1:]
+    result = sequential_sum_product(ops.logaddexp, ops.add,
```
Comment (fritzo) [inline examples/mixed_hmm/experiment.py]: nit: can we replace `sequential_sum_product` -> `MarkovProduct`?

File: examples/mixed_hmm/model.py [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+    Convert raw observation tensors into funsor.torch.Tensors
+    """
+    batch_inputs = OrderedDict([
+        ("i", bint(config["sizes"]["individual"])),
+        ("g", bint(config["sizes"]["group"])),
+        ("t", bint(config["sizes"]["timesteps"])),
+    ])
+
+    observations = {}
+    for name, data in config["observations"].items():
```
Comment (fritzo) [inline examples/mixed_hmm/model.py]: note: I sometimes use the terser notation
```py
T = config["sizes"]["timesteps"]
observations = {name: Tensor(data[..., :T])["i", "g", "t"]  # <--- getitem + str
                for name, data in config["observations"].items()}
```

File: funsor/distributions.py [hunk 6]
```diff
@@ -545,6 +545,48 @@ def eager_poisson(rate, value):
     return Poisson.eager_log_prob(rate=rate, value=value)
 
 
+class Gamma(Distribution):
```
Comment (fritzo) [inline funsor/distributions.py]: Could you factor these out as a separate PR and add simple tests?

File: examples/mixed_hmm/model.py [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+
+    N_state = config["sizes"]["state"]
+
+    params = initialize_model_params(config)
+    observations = initialize_observations(config)
+    raggedness_masks = initialize_raggedness_masks(config)
+    zi_masks = initialize_zi_masks(config)
+
+    # initialize gamma to uniform
+    gamma = Tensor(
```
Comment (fritzo) [inline examples/mixed_hmm/model.py]: note: ditto I often abbreviate this as
```py
gamma = Tensor(torch.zeros(N_state, N_state))["y_prev"]
```

File: examples/mixed_hmm/model.py [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+        torch.zeros((N_state, N_state)),
+        OrderedDict([("y_prev", bint(N_state))]),
+    )
+
+    N_v = config["sizes"]["random"]
+    N_c = config["sizes"]["group"]
+    log_prob = []
+
+    plate_g = Tensor(torch.zeros(N_c), OrderedDict([("g", bint(N_c))]))
```
Comment (fritzo) [inline examples/mixed_hmm/model.py]: note: This is an interesting pattern, denoting plates as zero `Tensor`s. I think we could avoid this after #233 , since plate size can be determined from context. (unrelated to this PR, but it would be nice cleanup in the future).

File: examples/mixed_hmm/model.py [hunk 9]
```diff
@@ -0,0 +1,527 @@
...
+            eps_i_dist = plate_g + plate_i + dist.Normal(**params["eps_i"])(value="eps_i")
+
+        log_prob += eps_i_dist
+
+    return log_prob
+
+
+def model_sequential(config):
+    """
+    Simpler version of generic model with no zero-inflation
```
Comment (fritzo) [inline examples/mixed_hmm/model.py]: note: if you do want to model zero-inflation, you could use `Stack`, e.g. as I've done in the BART model https://github.com/pyro-ppl/sandbox/blob/55ff040/2019-08-time-series/bart/forecast.py#L201
```py
likelihood = fdist.Categorical(gate, value="gated")
likelihood += funsor.Stack("gated", (
    fdist.Poisson(rate, value=data),
    fdist.Delta(0, value=data)))
likelihood = likelihood.reduce(ops.logaddexp, "gated")
```

File: examples/mixed_hmm/model.py [hunk 10]
```diff
@@ -0,0 +1,397 @@
...
+                value="zi_omega")
+            omega_zi_dist = plate_g + plate_i + dist.Delta(self.config["MISSING"])(
+                value=self.observations["omega"])
+            omega_dist = (omega_zi + Stack("zi_omega", (omega_dist, omega_zi_dist))).reduce(ops.logaddexp, "zi_omega")
+
+        # finally, construct the term for parallel scan reduction
+        hmm_factor = step_dist + angle_dist + omega_dist
+        hmm_factor = hmm_factor * self.raggedness_masks["individual"]
+        hmm_factor = hmm_factor * self.raggedness_masks["timestep"]
+        # copy masking behavior of pyro.infer.TraceEnum_ELBO._compute_model_factors
```
Comment (eb8680) [inline examples/mixed_hmm/model.py]: Compare to https://github.com/pyro-ppl/pyro/blob/dev/pyro/infer/traceenum_elbo.py#L95

**Comment by fritzo** [comment]
What is the status of this PR? Can we merge it soon?

File: funsor/distributions.py (hunk 6 above)
Comment (eb8680) [reply funsor/distributions.py]: Done in #268

**Comment by eb8680** [comment]
> What is the status of this PR? Can we merge it soon?

Should be good to merge after #268
